Graceful Startup Issue - liberty-kafka connector: @Incoming methods receive messages before application startup completes  — causes permanent channel death

## Describe the bug

The `liberty-kafka` connector begins delivering messages to `@Incoming` methods before the application is fully started. When the `@Incoming` bean injects an EJB `@Startup @Singleton` (via `@Inject`), and Kafka has pending messages at startup, the consumer method is invoked before the EJB is registered in JNDI. This causes a `WELD-000079` error that **permanently kills** the internal reactive messaging channel pipeline. No further messages are ever processed — the consumer silently fails with no recovery mechanism short of a pod restart.

After 5 minutes of inactivity (`max.poll.interval.ms`), the Kafka broker evicts the consumer from the group entirely. The partition becomes unassigned and messages accumulate permanently. The application appears healthy (liveness UP, no alerts).

**Full stack trace:**

```
[ERROR] SRMSG00200: The method MyKafkaConsumer#consume has thrown an exception
io.smallrye.reactive.messaging.providers.ProcessingException: SRMSG00103: Exception thrown when calling the method MyKafkaConsumer#consume
    at io.smallrye.reactive.messaging.providers.AbstractMediator.invoke(AbstractMediator.java:162)
    at io.smallrye.reactive.messaging.providers.AbstractMediator.lambda$invokeOnMessageContext$13(AbstractMediator.java:168)
    at io.smallrye.reactive.messaging.providers.locals.LocalContextMetadata.lambda$invokeOnMessageContext$0(LocalContextMetadata.java:34)
    at io.smallrye.reactive.messaging.providers.locals.LocalContextMetadata.lambda$invokeOnMessageContext$3(LocalContextMetadata.java:61)
    at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
    at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
    at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
    at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
    at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap.subscribe(UniOnItemOrFailureFlatMap.java:27)
    at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
    at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.request(UniToMultiPublisher.java:74)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber.innerOnSubscribe(MultiConcatMapOp.java:109)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber$InnerSubscriber.onSubscribe(MultiConcatMapOp.java:297)
    at io.smallrye.mutiny.converters.uni.UniToMultiPublisher.subscribe(UniToMultiPublisher.java:25)
    at io.smallrye.mutiny.groups.MultiCreate$1.subscribe(MultiCreate.java:165)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber.onItem(MultiConcatMapOp.java:119)
    at io.smallrye.mutiny.operators.multi.MultiMapOp$MapProcessor.onItem(MultiMapOp.java:50)
    at io.smallrye.mutiny.operators.multi.MultiOnItemInvoke$MultiOnItemInvokeProcessor.onItem(MultiOnItemInvoke.java:41)
    at io.smallrye.mutiny.operators.multi.MultiOperatorProcessor.onItem(MultiOperatorProcessor.java:100)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber.innerOnItem(MultiConcatMapOp.java:133)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber$InnerSubscriber.onItem(MultiConcatMapOp.java:302)
    at io.smallrye.mutiny.operators.multi.builders.IterableBasedMulti$IteratorSubscription.slowPath(IterableBasedMulti.java:160)
    at io.smallrye.mutiny.operators.multi.builders.IterableBasedMulti$IteratorSubscription.request(IterableBasedMulti.java:71)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber.innerOnSubscribe(MultiConcatMapOp.java:109)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber$InnerSubscriber.onSubscribe(MultiConcatMapOp.java:297)
    at io.smallrye.mutiny.operators.multi.builders.IterableBasedMulti.subscribe(IterableBasedMulti.java:48)
    at io.smallrye.mutiny.operators.AbstractMulti.subscribe(AbstractMulti.java:60)
    at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber.onItem(MultiConcatMapOp.java:119)
    at io.smallrye.mutiny.operators.multi.MultiSelectFirstWhileOp$MultiSelectFirstWhileProcessor.onItem(MultiSelectFirstWhileOp.java:59)
    at io.smallrye.mutiny.operators.multi.MultiOnItemInvoke$MultiOnItemInvokeProcessor.onItem(MultiOnItemInvoke.java:41)
Caused by: org.jboss.weld.exceptions.CreationException: WELD-000079: Could not instantiate a proxy for a session bean:
  Session bean [class MyExecutorService with qualifiers [@Any @Default]; local interfaces are [MyExecutorService]]
    at org.jboss.weld.bean.proxy.ProxyFactory.create(ProxyFactory.java:370)
    at org.jboss.weld.bean.SessionBean.createClientProxy(SessionBean.java:230)
    at org.jboss.weld.bean.SessionBean.getClientProxy(SessionBean.java:240)
    at org.jboss.weld.bean.builtin.InstanceImpl.getBeanInstance(InstanceImpl.java:242)
    at org.jboss.weld.bean.builtin.InstanceImpl.get(InstanceImpl.java:114)
    at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
```

**5 minutes later — consumer silently evicted:**
```
[WARN] consumer poll timeout has expired. This means the time between subsequent
  calls to poll() was longer than the configured max.poll.interval.ms
[WARN] Member sending LeaveGroup request to coordinator due to consumer poll timeout has expired.
[WARN] Resetting generation and member id due to: consumer pro-actively leaving the group
```

---

## Steps to Reproduce

1. Create an EJB `@Startup @Singleton` that depends on another singleton (simulating a real initialization chain):

```java
@Singleton
@Startup
@DependsOn("DatabaseInitializer")
public class MyExecutorService {
    private ExecutorService executor;

    @PostConstruct
    public void init() {
        // Takes ~1-2 seconds due to @DependsOn chain
        executor = Executors.newFixedThreadPool(10);
    }

    public void execute(Runnable task) {
        executor.execute(task);
    }
}
```

2. Create an `@ApplicationScoped` CDI bean with `@Incoming` that injects the EJB:

```java
@ApplicationScoped
public class MyKafkaConsumer {

    @Inject
    private MyExecutorService executorService;

    @Incoming("my-channel")
    @Transactional(Transactional.TxType.REQUIRED)
    public CompletionStage<Void> consume(Message<String> message) {
        executorService.execute(() -> {
            // Process message
        });
        return message.ack();
    }
}
```

3. Configure the channel in `microprofile-config.properties`:

```properties
mp.messaging.incoming.my-channel.connector=liberty-kafka
mp.messaging.incoming.my-channel.topic=my-topic
mp.messaging.incoming.my-channel.group.id=my-group
mp.messaging.incoming.my-channel.key.deserializer=org.apache.kafka.common.serialization.StringDeserializer
mp.messaging.incoming.my-channel.value.deserializer=org.apache.kafka.common.serialization.StringDeserializer
```

4. **Produce one or more messages to `my-topic` BEFORE starting the server**

5. Start the Liberty server

6. Observe `SRMSG00200` + `WELD-000079` in `messages.log` — channel dies permanently, no further messages processed

**Important:** The bug only reproduces when Kafka has pending messages at the moment the application starts. If the topic is empty at startup, no error occurs (but the application remains vulnerable to the race on subsequent restarts).

---

## Expected behavior

Messages should not be delivered to `@Incoming` methods until the application is fully started and all injectable dependencies (including EJB singletons) are resolvable. The connector should either:

1. Wait for the application `STARTED` lifecycle event before beginning to deliver messages to `@Incoming` methods, OR
2. Provide a configuration property to control when consumption begins, OR
3. Expose a programmatic pause/resume mechanism on channels (similar to SmallRye's `PausableChannel`)

---

## Diagnostic information

- **OpenLiberty Version:** 26.0.0.1 (also reproduced on 24.0.0.x and 25.0.0.x)
- **Affected feature(s):** `mpReactiveMessaging-3.0`, `enterpriseBeans-4.0`, `cdi-4.0`
- **Java Version:**
  ```
  openjdk version "17.0.x" 2024-xx-xx
  OpenJDK Runtime Environment Semeru (build 17.0.x+x)
  IBM Semeru Runtime Open Edition (build 17.0.x+x, mixed mode, sharing)
  ```
- **server.xml configuration:**

```xml
<server description="Reproduction server">
    <featureManager>
        <feature>mpReactiveMessaging-3.0</feature>
        <feature>enterpriseBeans-4.0</feature>
        <feature>cdi-4.0</feature>
        <feature>concurrent-3.0</feature>
        <feature>mpHealth-4.0</feature>
    </featureManager>

    <httpEndpoint id="defaultHttpEndpoint" httpPort="9080" httpsPort="9443"/>
</server>
```

---

## Additional context

### Root cause (from source code inspection)

The `liberty-kafka` connector subscribes and begins polling during the CDI `@AfterDeploymentValidation` lifecycle event (`OLReactiveMessaging30Extension.afterDeploymentValidation()` → `startMediatorManager()` → `KafkaIncomingConnector.getPublisherBuilder()` → `KafkaInput.pollKafkaAsync()`).

This occurs **before** the EJB container finishes processing `@Startup @Singleton @DependsOn` chains, and **before** `SimpleDeployedAppInfoBase.postDeployApp()` fires `stateChangeService.fireApplicationStarted()`.

```
T+0ms      CDI @AfterDeploymentValidation → MediatorManager.start()
           → KafkaConsumer subscribes and polls immediately
T+200ms    Kafka delivers pending message → @Incoming method invoked
           → @Inject of EJB → JNDI lookup fails → WELD-000079
           → Channel pipeline STOPS PERMANENTLY
T+1500ms   EJB @Startup @DependsOn chain completes (too late)
T+5min     max.poll.interval.ms expires → consumer evicted from group
```

### What we investigated (all exhausted)

| Approach | Result |
|----------|--------|
| SmallRye `PausableChannel` + `ChannelRegistry` | `WELD-001408: Unsatisfied dependencies` — not available on Liberty |
| `liberty-kafka` connector config properties | Only `topic`, `unacked.limit`, `fast.ack`, `context.service`, `creation.retry.seconds` exist. None delay consumption. |
| JMX `ServerEndpointControl` pause/resume | Only for MDB/JCA endpoints, not MP RM channels |
| `@DependsOn` on the `@Incoming` bean | CDI beans cannot use `@DependsOn` (EJB-only annotation) |
| `@Singleton @Startup` on the consumer | `@Incoming` wiring is CDI-driven, ignores EJB annotations |
| `@RequestScoped` / `@Dependent` | MP RM spec requires `@ApplicationScoped` for `@Incoming` beans |
| CDI `@Priority` | Does not affect reactive stream subscription timing |
| `@Observes @Initialized(ApplicationScoped.class)` | Fires before `@AfterDeploymentValidation` — cannot prevent subscription |
| Liberty `server pause` command | HTTP + JCA only, not MP RM channels |
| Source code inspection (`KafkaInput.java`, `KafkaIncomingConnector.java`, `OLReactiveMessaging30Extension.java`) | Confirmed no readiness gate exists |

### Why this is critical in production

- Reproduces **100% of the time** during K8s rolling deployments when Kafka has pending messages
- **Silent failure** — pod liveness/readiness probes remain UP, no alerting fires
- **Permanent message loss** — partition becomes unassigned, lag grows indefinitely
- **No programmatic recovery** — channel cannot be restarted without restarting the JVM

### Current workaround

Using `Instance<T>` (lazy EJB resolution) + a `CountDownLatch` gate that blocks the consumer thread until the application signals readiness. This keeps the channel alive (blocking rather than throwing) and relies on `unacked.limit` back-pressure to prevent further polling while blocked.

### Related issues

- #12910 — Can't use mpReactiveMessaging with an EJB (same WELD-000079, different scenario)
- #10937 — Kafka connector does not retry connection (resolved with `creation.retry.seconds`, does not address startup ordering)

Hello @SheikhSuhail19 

Thank you for your detailed report. I wasn't able to recreate it locally but I _was_ able to verify that we were requesting the async action to poll for new messages before EJB was registering beans with JNDI. Perhaps an environmental difference, especially as I built my recreate into our test framework to live there in the future.

That is enough for me to be confident this is a bug and I am working on a fix.

-------

I do have one burning question though. 

```
Caused by: org.jboss.weld.exceptions.CreationException: WELD-000079: Could not instantiate a proxy for a session bean:
  Session bean [class MyExecutorService with qualifiers [@Any @Default]; local interfaces are [MyExecutorService]]
    at org.jboss.weld.bean.proxy.ProxyFactory.create(ProxyFactory.java:370)
    at org.jboss.weld.bean.SessionBean.createClientProxy(SessionBean.java:230)
    at org.jboss.weld.bean.SessionBean.getClientProxy(SessionBean.java:240)
    at org.jboss.weld.bean.builtin.InstanceImpl.getBeanInstance(InstanceImpl.java:242)
    at org.jboss.weld.bean.builtin.InstanceImpl.get(InstanceImpl.java:114)
    at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
```
How did you get this stack? I can't find _any_ version of weld where `FieldInjectionPoint.inject()` calls a method called `get()`? If you are using `cdi-4.0` you should be using weld 5.1.1.SP2, and [as you can see](https://github.com/weld/core/blob/5.1.1.SP2/impl/src/main/java/org/jboss/weld/injection/FieldInjectionPoint.java#L92) line 92 calls `getInjectableReference()` not `get()`. Are you running your own patched weld jar?

> Hello [@SheikhSuhail19](https://github.com/SheikhSuhail19)
> 
> Thank you for your detailed report. I wasn't able to recreate it locally but I _was_ able to verify that we were requesting the async action to poll for new messages before EJB was registering beans with JNDI. Perhaps an environmental difference, especially as I built my recreate into our test framework to live there in the future.
> 
> That is enough for me to be confident this is a bug and I am working on a fix.
> 
> I do have one burning question though.
> 
> ```
> Caused by: org.jboss.weld.exceptions.CreationException: WELD-000079: Could not instantiate a proxy for a session bean:
>   Session bean [class MyExecutorService with qualifiers [@Any @Default]; local interfaces are [MyExecutorService]]
>     at org.jboss.weld.bean.proxy.ProxyFactory.create(ProxyFactory.java:370)
>     at org.jboss.weld.bean.SessionBean.createClientProxy(SessionBean.java:230)
>     at org.jboss.weld.bean.SessionBean.getClientProxy(SessionBean.java:240)
>     at org.jboss.weld.bean.builtin.InstanceImpl.getBeanInstance(InstanceImpl.java:242)
>     at org.jboss.weld.bean.builtin.InstanceImpl.get(InstanceImpl.java:114)
>     at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
> ```
> 
> How did you get this stack? I can't find _any_ version of weld where `FieldInjectionPoint.inject()` calls a method called `get()`? If you are using `cdi-4.0` you should be using weld 5.1.1.SP2, and [as you can see](https://github.com/weld/core/blob/5.1.1.SP2/impl/src/main/java/org/jboss/weld/injection/FieldInjectionPoint.java#L92) line 92 calls `getInjectableReference()` not `get()`. Are you running your own patched weld jar?


Hi @benjamin-confino,

Thank you for confirming the bug and for the thorough investigation!

Regarding your question about the stack trace — you're absolutely right to call that out. The original logs are no longer accessible to me, earlier I reconstructed and anonymised it when writing the issue. In doing so I likely introduced inaccuracies, so please disregard that specific Caused by block. The core behaviour (WELD-000079 at startup → permanent channel death) is what we actually experienced.

Really glad the fix is in progress — this one has been painful issue and I'm sure other teams running Liberty with EJBs and reactive messaging will benefit from it too. We did manage a programmatic workaround in the meantime (using Instance<T> for lazy EJB resolution combined with a CountDownLatch gate to block the consumer until the application signals readiness), but a proper framework-level fix is obviously the right solution.

--- a/dev/com.ibm.ws.microprofile.reactive.messaging.kafka/bnd.bnd
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging.kafka/bnd.bnd
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2019, 2023 IBM Corporation and others.
+# Copyright (c) 2019, 2026 IBM Corporation and others.
 # All rights reserved. This program and the accompanying materials
 # are made available under the terms of the Eclipse Public License 2.0
 # which accompanies this distribution, and is available at
@@ -26,7 +26,8 @@ WS-TraceGroup: REACTIVEMESSAGE
 -cdiannotations:
 
 -dsannotations-inherit: true
--dsannotations: com.ibm.ws.microprofile.reactive.messaging.kafka.KafkaConnectorCDIExtension
+-dsannotations: com.ibm.ws.microprofile.reactive.messaging.kafka.KafkaConnectorCDIExtension,\
+ com.ibm.ws.microprofile.reactive.messaging.kafka.KafkaApplicationStateListener
 
 Export-Package: com.ibm.ws.microprofile.reactive.messaging.kafka,\
                 com.ibm.ws.kafka.security,\
@@ -41,25 +42,30 @@ Include-Resource: \
 
 
 -buildpath: \
-  com.ibm.ws.microprofile.reactive.messaging.kafka.adapter;version=latest, \
-  io.openliberty.microprofile.reactive.messaging.internal;version=latest, \
-  com.ibm.ws.logging;version=latest, \
-  com.ibm.ws.logging.core;version=latest, \
-  org.eclipse.osgi;version=latest, \
-  com.ibm.websphere.javaee.cdi.2.0;version=latest, \
-  com.ibm.websphere.org.osgi.core;version=latest, \
-  com.ibm.websphere.org.osgi.service.component;version=latest, \
-  com.ibm.wsspi.org.osgi.service.component.annotations;version=latest, \
-  com.ibm.websphere.org.eclipse.microprofile.reactive.streams.operators.1.0;version=latest, \
-  com.ibm.websphere.org.reactivestreams.reactive-streams.1.0;version=latest, \
-  com.ibm.websphere.org.eclipse.microprofile.reactive.messaging.1.0;version=latest, \
-  com.ibm.websphere.org.eclipse.microprofile.config.1.3;version=latest, \
-  com.ibm.ws.org.osgi.annotation.versioning;version=latest, \
-  com.ibm.websphere.javaee.annotation.1.2;version=latest, \
-  com.ibm.websphere.javaee.concurrent.1.0;version=latest, \
-  com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.impl;version=latest, \
-  com.ibm.ws.cdi.interfaces;version=latest, \
-  com.ibm.websphere.appserver.api.passwordUtil;version=latest
+	com.ibm.ws.microprofile.reactive.messaging.kafka.adapter;version=latest,\
+	io.openliberty.microprofile.reactive.messaging.internal;version=latest,\
+	com.ibm.ws.logging;version=latest,\
+	com.ibm.ws.logging.core;version=latest,\
+	org.eclipse.osgi;version=latest,\
+	com.ibm.websphere.javaee.cdi.2.0;version=latest,\
+	com.ibm.websphere.org.osgi.core;version=latest,\
+	com.ibm.websphere.org.osgi.service.component;version=latest,\
+	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.reactive.streams.operators.1.0;version=latest,\
+	com.ibm.websphere.org.reactivestreams.reactive-streams.1.0;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.reactive.messaging.1.0;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.config.1.3;version=latest,\
+	com.ibm.ws.org.osgi.annotation.versioning;version=latest,\
+	com.ibm.websphere.javaee.annotation.1.2;version=latest,\
+	com.ibm.websphere.javaee.concurrent.1.0;version=latest,\
+	com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.impl;version=latest,\
+	com.ibm.ws.cdi.interfaces;version=latest,\
+	com.ibm.websphere.appserver.api.passwordUtil;version=latest,\
+	com.ibm.ws.app.manager;version=latest,\
+	com.ibm.ws.container.service.compat;version=latest,\
+	com.ibm.ws.kernel.service;version=latest,\
+	com.ibm.ws.container.service;version=latest,\
+	com.ibm.ws.app.manager;version=latest
   
 -testpath: \
   org.hamcrest:hamcrest-all;version=1.3, \

--- a/dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/KafkaApplicationStateListener.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/KafkaApplicationStateListener.java
@@ -1,0 +1,79 @@
+/*******************************************************************************
+ * Copyright (c) 2026 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.kafka;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+
+import com.ibm.ws.container.service.app.deploy.ApplicationInfo;
+import com.ibm.ws.container.service.state.ApplicationStateListener;
+import com.ibm.ws.container.service.state.StateChangeException;
+
+/**
+ * An OSGi component that listens to application lifecycle events and tracks
+ * each application's started state using a {@link CountDownLatch}.
+ * <p>
+ * On {@code applicationStarting} a latch with count 1 is placed in the map.
+ * On {@code applicationStarted} the latch is counted down to zero, unblocking
+ * any thread waiting on it. On {@code applicationStopped} the entry is removed
+ * from the map.
+ *
+ * This class tolerates gets being called before application starting
+ */
+@Component(service = { KafkaApplicationStateListener.class,
+                       ApplicationStateListener.class }, configurationPolicy = ConfigurationPolicy.IGNORE, immediate = true)
+public class KafkaApplicationStateListener implements ApplicationStateListener {
+
+    private final ConcurrentHashMap<String, CountDownLatch> appLatches = new ConcurrentHashMap<>();
+
+    /** {@inheritDoc} */
+    @Override
+    public void applicationStarting(ApplicationInfo appInfo) throws StateChangeException {
+        appLatches.computeIfAbsent(appInfo.getDeploymentName(), key -> new CountDownLatch(1)); //Tolerate gets being called before application starting.
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void applicationStarted(ApplicationInfo appInfo) throws StateChangeException {
+
+        CountDownLatch latch = appLatches.get(appInfo.getDeploymentName());
+        if (latch != null) {
+            latch.countDown();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void applicationStopping(ApplicationInfo appInfo) {
+        // Nothing to do on stopping
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void applicationStopped(ApplicationInfo appInfo) {
+        appLatches.remove(appInfo.getDeploymentName());
+    }
+
+    /**
+     * Returns the {@link CountDownLatch} associated with the named application
+     *
+     * @param appName the application name
+     * @return the latch
+     */
+    public CountDownLatch getLatch(String appName) {
+        return appLatches.computeIfAbsent(appName, key -> new CountDownLatch(1)); //Tolerate gets being called before application starting.
+    }
+}

--- a/dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/KafkaInput.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/KafkaInput.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2019, 2023 IBM Corporation and others.
+ * Copyright (c) 2019, 2026 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -24,6 +24,7 @@ import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Function;
@@ -34,8 +35,10 @@ import org.eclipse.microprofile.reactive.streams.operators.ReactiveStreams;
 
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.cdi.CDIService;
 import com.ibm.ws.ffdc.FFDCFilter;
 import com.ibm.ws.ffdc.annotation.FFDCIgnore;
+import com.ibm.ws.kernel.service.util.ServiceCaller;
 import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.ConsumerRebalanceListener;
 import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.ConsumerRecord;
 import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.ConsumerRecords;
@@ -44,6 +47,9 @@ import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.KafkaConsumer;
 import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.OffsetAndMetadata;
 import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.TopicPartition;
 import com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.WakeupException;
+import com.ibm.ws.runtime.metadata.ComponentMetaData;
+import com.ibm.ws.threadContext.ComponentMetaDataAccessorImpl;
+import com.ibm.wsspi.kernel.service.utils.FrameworkState;
 
 import io.openliberty.microprofile.reactive.messaging.internal.interfaces.RMAsyncProvider;
 import io.openliberty.microprofile.reactive.messaging.internal.interfaces.RMContext;
@@ -64,6 +70,28 @@ public class KafkaInput<K, V> implements ConsumerRebalanceListener {
     private final KafkaConsumer<K, V> kafkaConsumer;
     private final RMAsyncProvider asyncProvider;
     private final Collection<String> topics;
+
+    /**
+     * The name of the application that created this KafkaInput, captured at construction time.
+     * May be {@code null} if the application name could not be determined.
+     */
+    private final String applicationName;
+
+    /**
+     * Cached flag set to {@code true} once the application-started latch has been observed to have
+     * counted down. Once {@code true}, subsequent calls to {@link #executePollActions} skip the
+     * latch await entirely for performance.
+     */
+    private volatile boolean applicationStarted = false;
+
+    /**
+     * Latch obtained from {@link KafkaApplicationStateListener} for the owning application.
+     * Counted down to zero once the application has fully started.
+     * {@link #executePollActions} awaits this latch before entering its poll loop so
+     * that messages are never dispatched to application code before the application
+     * is fully started.
+     */
+    private final CountDownLatch applicationStartedLatch;
 
     private PublisherBuilder<Message<V>> publisher;
     private boolean subscribed = false;
@@ -100,6 +128,9 @@ public class KafkaInput<K, V> implements ConsumerRebalanceListener {
      */
     private final ReentrantLock lock = new ReentrantLock();
 
+    private final static ServiceCaller<KafkaApplicationStateListener> APP_STATE_LISTENER_CALLER = new ServiceCaller<>(KafkaInput.class,
+                                                                                                                      KafkaApplicationStateListener.class);
+
     public KafkaInput(KafkaAdapterFactory kafkaAdapterFactory, PartitionTrackerFactory partitionTrackerFactory,
                       KafkaConsumer<K, V> kafkaConsumer, RMAsyncProvider asyncProvider,
                       String topic, int unackedLimit, boolean fastAck) {
@@ -116,6 +147,40 @@ public class KafkaInput<K, V> implements ConsumerRebalanceListener {
             this.unackedMessageCounter = ThresholdCounter.UNLIMITED;
         }
         this.fastAck = fastAck;
+        this.applicationName = resolveApplicationName();
+
+        if (applicationName == null) {
+            throw new IllegalStateException("Could not determine the application name for this KafkaInput");
+        }
+
+        this.applicationStartedLatch = APP_STATE_LISTENER_CALLER.run(kasl -> kasl.getLatch(applicationName))
+                                                                .orElseThrow(() -> new IllegalStateException("Could not acquire the latch KafkaInput with applicationName "
+                                                                                                             + applicationName));
+
+    }
+
+    /**
+     * Capture the application name at construction time.
+     * <p>
+     * Mirrors OSGiConfigUtils.getApplicationName (in io.openliberty.microprofile.config.internal.serverxml):
+     * reads the thread-local {@link ComponentMetaData} first then falls back to asking the {@link CDIService}
+     * for the current application context ID.
+     *
+     * @return the application name, or {@code null} if it cannot be determined
+     */
+    private static String resolveApplicationName() {
+        if (!FrameworkState.isValid()) {
+            return null;
+        }
+        ComponentMetaData cmd = ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor().getComponentMetaData();
+        if (cmd != null) {
+            return cmd.getJ2EEName().getApplication();
+        }
+
+        //Fallback to try getting the name from CDI. This will work if CDI's startup routine is on the current thread stack.
+        return ServiceCaller.runOnce(KafkaInput.class, CDIService.class,
+                                     CDIService::getCurrentApplicationContextID)
+                            .orElse(null);
     }
 
     public PublisherBuilder<Message<V>> getPublisher() {
@@ -198,6 +263,10 @@ public class KafkaInput<K, V> implements ConsumerRebalanceListener {
         }
 
         this.running = false;
+        // Unlock the application-started latch so threads
+        // can observe running == false and exit cleanly.
+        applicationStartedLatch.countDown();
+
         this.kafkaConsumer.wakeup();
         this.lock.lock();
         try {
@@ -213,6 +282,7 @@ public class KafkaInput<K, V> implements ConsumerRebalanceListener {
 
     @FFDCIgnore({ WakeupException.class, RejectedExecutionException.class })
     private CompletionStage<PublisherBuilder<Message<V>>> pollKafkaAsync() {
+
         if (!this.subscribed) {
             this.lock.lock();
             try {
@@ -235,7 +305,7 @@ public class KafkaInput<K, V> implements ConsumerRebalanceListener {
 
         ConsumerRecords<K, V> records = null;
 
-        while (this.lock.tryLock()) {
+        while (applicationStarted && this.lock.tryLock()) {
             try {
                 records = this.kafkaConsumer.poll(ZERO);
                 break;
@@ -271,9 +341,22 @@ public class KafkaInput<K, V> implements ConsumerRebalanceListener {
 
     /**
      * Run any pending actions and then poll Kafka for messages.
+     * <p>
+     * Waits until the owning application has fully started before entering the poll loop,
+     * so that messages are never dispatched to application code during application startup.
+     * The wait is skipped on subsequent calls once {@link #applicationStarted} is {@code true}.
      */
-    @FFDCIgnore(WakeupException.class)
+    @FFDCIgnore({ WakeupException.class, InterruptedException.class })
     private void executePollActions(CompletableFuture<PublisherBuilder<Message<V>>> result) {
+        if (!applicationStarted) {
+            try {
+                applicationStartedLatch.await();
+                applicationStarted = true;
+            } catch (InterruptedException e) {
+                result.completeExceptionally(e);
+                return;
+            }
+        }
         this.lock.lock();
         try {
             while (this.running) {

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/bnd.bnd
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/bnd.bnd
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2019, 2023 IBM Corporation and others.
+# Copyright (c) 2019, 2026 IBM Corporation and others.
 # All rights reserved. This program and the accompanying materials
 # are made available under the terms of the Eclipse Public License 2.0
 # which accompanies this distribution, and is available at
@@ -42,20 +42,25 @@ tested.features:\
   restfulwsclient-4.0,\
   restfulws-3.0, \
   restfulws-4.0, \
-  jsonp-2.0
+  jsonp-2.0, \
+  connectors-2.1, \
+  concurrent-3.1
 
 -buildpath: \
-    io.openliberty.microprofile.reactive.messaging.internal_fat.common.jakarta;version=latest,\
-    io.openliberty.org.eclipse.microprofile.reactive.messaging.3.0;version=latest,\
-    io.openliberty.org.eclipse.microprofile.reactive.streams.operators.3.0;version=latest,\
-    com.ibm.websphere.org.reactivestreams.reactive-streams.1.0;version=latest,\
-    io.openliberty.org.eclipse.microprofile.config.3.0;version=latest,\
-    io.openliberty.jakarta.cdi.3.0;version=latest,\
-    io.openliberty.jakarta.servlet.5.0;version=latest,\
-    io.openliberty.jakarta.restfulWS.3.0;version=latest,\
-    org.apache.kafka:kafka-clients;version=3.5.1,\
-    io.openliberty.org.testcontainers;version=latest,\
-    com.ibm.ws.componenttest.2.0;version=latest
+	io.openliberty.microprofile.reactive.messaging.internal_fat.common.jakarta;version=latest,\
+	io.openliberty.org.eclipse.microprofile.reactive.messaging.3.0;version=latest,\
+	io.openliberty.org.eclipse.microprofile.reactive.streams.operators.3.0;version=latest,\
+	com.ibm.websphere.org.reactivestreams.reactive-streams.1.0;version=latest,\
+	io.openliberty.org.eclipse.microprofile.config.3.0;version=latest,\
+	io.openliberty.jakarta.cdi.3.0;version=latest,\
+	io.openliberty.jakarta.servlet.5.0;version=latest,\
+	io.openliberty.jakarta.restfulWS.3.0;version=latest,\
+	org.apache.kafka:kafka-clients;version='3.5.1',\
+	io.openliberty.org.testcontainers;version=latest,\
+	com.ibm.ws.componenttest.2.0;version=latest,\
+	io.openliberty.jakarta.enterpriseBeans.4.0;version=latest,\
+	io.openliberty.jakarta.annotation.2.0;version=latest,\
+	io.openliberty.jakarta.transaction.2.0;version=latest
 
 -dependson.1: \
     com.ibm.ws.org.slf4j.jdk14

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/ExampleDependency.java
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/ExampleDependency.java
@@ -1,0 +1,20 @@
+/*******************************************************************************
+ * Copyright (c) 2026 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *******************************************************************************/
+package io.openliberty.microprofile.reactive.messaging.fat.apps.startup;
+
+import jakarta.ejb.Singleton;
+import jakarta.ejb.Startup;
+
+@Singleton
+@Startup
+public class ExampleDependency {
+
+}

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/TestMessageOnStartupKafkaConsumer.java
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/TestMessageOnStartupKafkaConsumer.java
@@ -1,0 +1,50 @@
+/*******************************************************************************
+ * Copyright (c) 2026 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *******************************************************************************/
+package io.openliberty.microprofile.reactive.messaging.fat.apps.startup;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractReceptionBean;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+@ApplicationScoped
+public class TestMessageOnStartupKafkaConsumer extends AbstractReceptionBean<String> {
+
+    @Incoming("TestMessageOnStartup")
+    @Override
+    public CompletionStage<Void> receiveMessage(Message<String> message) {
+        System.out.println("Message recievd " + message.getPayload());
+
+        InitialContext ctx;
+        try {
+            ctx = new InitialContext();
+
+            ExampleDependency exampleDependency = (ExampleDependency) ctx
+                            .lookup("java:module/ExampleDependency");
+
+            exampleDependency.toString();
+
+        } catch (NamingException e) {
+            throw new RuntimeException("Failed to lookup the EJB bean " + e);
+        }
+
+        return super.receiveMessage(message);
+    }
+
+}

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/TestMessageOnStartupServlet.java
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/TestMessageOnStartupServlet.java
@@ -1,0 +1,52 @@
+/*******************************************************************************
+ * Copyright (c) 2026 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *******************************************************************************/
+package io.openliberty.microprofile.reactive.messaging.fat.apps.startup;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.junit.Test;
+
+import componenttest.app.FATServlet;
+import jakarta.inject.Inject;
+import jakarta.servlet.annotation.WebServlet;
+
+@SuppressWarnings("serial")
+@WebServlet("/TestMessageOnStartup")
+public class TestMessageOnStartupServlet extends FATServlet {
+
+    private static final Duration TIMEOUT = Duration.ofSeconds(30);
+
+    // The five messages sent to Kafka before the server started (see TestMessageOnStartupTest.setup())
+    private static final int EXPECTED_MESSAGE_COUNT = 5;
+
+    @Inject
+    private TestMessageOnStartupKafkaConsumer consumer;
+
+    @Test
+    public void testPreloadedMessagesWorkAfterStartup() throws Exception {
+
+        List<Message<String>> received = consumer.assertReceivedMessages(EXPECTED_MESSAGE_COUNT, TIMEOUT);
+        List<String> payloads = received.stream().map(Message::getPayload).collect(Collectors.toList());
+        assertThat(payloads, containsInAnyOrder(
+                                                "test message 1",
+                                                "test message 2",
+                                                "test message 3",
+                                                "test message 4",
+                                                "test message 5"));
+    }
+
+}

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/extension/DelayingCDIExtension.java
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/apps/startup/extension/DelayingCDIExtension.java
@@ -1,0 +1,36 @@
+/*******************************************************************************
+ * Copyright (c) 2026 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+ package io.openliberty.microprofile.reactive.messaging.fat.apps.startup.extension;
+
+import jakarta.annotation.Priority; 
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.inject.spi.AfterDeploymentValidation;
+import jakarta.enterprise.inject.spi.Extension;
+import jakarta.enterprise.inject.spi.BeanManager;
+
+public class DelayingCDIExtension implements Extension {
+
+    protected void afterDeploymentValidation(@Priority(Integer.MAX_VALUE) @Observes AfterDeploymentValidation done, BeanManager beanManager) {
+        //Delay until kafka messagse emitted before app startup risk being lost
+        System.out.println("afterDeploymentValidation - beginning delay");
+        
+        try {
+            Thread.sleep(3000);
+        }
+        catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    
+    }
+
+}

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/startup/TestMessageOnStartupTest.java
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/startup/TestMessageOnStartupTest.java
@@ -1,0 +1,120 @@
+/*******************************************************************************
+ * Copyright (c) 2026 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *******************************************************************************/
+package io.openliberty.microprofile.reactive.messaging.fat.startup;
+
+import static com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions.SERVER_ONLY;
+import static com.ibm.ws.microprofile.reactive.messaging.fat.kafka.common.ConnectorProperties.simpleIncomingChannel;
+import static com.ibm.ws.microprofile.reactive.messaging.fat.kafka.common.KafkaUtils.kafkaClientLibs;
+
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.EnterpriseArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.runner.RunWith;
+
+import com.ibm.websphere.simplicity.CDIArchiveHelper;
+import com.ibm.websphere.simplicity.PropertiesAsset;
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions;
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.common.ConnectorProperties;
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractReceptionBean;
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.KafkaTestClient;
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.KafkaWriter;
+import com.ibm.ws.microprofile.reactive.messaging.fat.repeats.ReactiveMessagingActions;
+
+import componenttest.annotation.Server;
+import componenttest.annotation.TestServlet;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.rules.repeater.RepeatTests;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.FATServletClient;
+import io.openliberty.microprofile.reactive.messaging.fat.apps.startup.TestMessageOnStartupServlet;
+import io.openliberty.microprofile.reactive.messaging.fat.suite.KafkaTests;
+
+@RunWith(FATRunner.class)
+@Mode(TestMode.FULL)
+public class TestMessageOnStartupTest extends FATServletClient {
+
+    public static final String APP_NAME = "TestMessageOnStartup";
+    public static final String SERVER_NAME = "TestMessageOnStartupServer";
+    public static final String TOPIC_NAME = APP_NAME;
+
+    @Server(SERVER_NAME)
+    @TestServlet(contextRoot = APP_NAME, servlet = TestMessageOnStartupServlet.class)
+    public static LibertyServer server;
+
+    @ClassRule
+    public static final RepeatTests r = ReactiveMessagingActions.reactive30Repeats(SERVER_NAME);
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        // Send five messages to the topic before the server starts, so they are
+        // waiting in the queue when MyKafkaConsumer.consume begins processing.
+        KafkaTestClient kafkaTestClient = new KafkaTestClient(KafkaTests.kafkaContainer.getBootstrapServers());
+        try (KafkaWriter<String, String> writer = kafkaTestClient.writerFor(TOPIC_NAME)) {
+            for (int i = 1; i <= 5; i++) {
+                writer.sendMessage("test message " + i);
+            }
+        }
+
+        ConnectorProperties inputConfig = simpleIncomingChannel(KafkaTests.connectionProperties(),
+                                                                APP_NAME,
+                                                                APP_NAME);
+
+        PropertiesAsset appConfig = new PropertiesAsset()
+                        .include(inputConfig);
+
+        WebArchive war = ShrinkWrap.create(WebArchive.class, APP_NAME + ".war")
+                        .addAsLibraries(kafkaClientLibs())
+                        .addPackage(TestMessageOnStartupServlet.class.getPackage())
+                        .addPackage(AbstractReceptionBean.class.getPackage())
+                        .addAsResource(appConfig, "META-INF/microprofile-config.properties");
+                        
+        JavaArchive jar = ShrinkWrap.create(JavaArchive.class, APP_NAME + ".jar")
+                        .addPackage(io.openliberty.microprofile.reactive.messaging.fat.apps.startup.extension.DelayingCDIExtension.class.getPackage());
+
+        CDIArchiveHelper.addJakartaCDIExtensionService(jar, io.openliberty.microprofile.reactive.messaging.fat.apps.startup.extension.DelayingCDIExtension.class);
+
+        String applicationXml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+                                "<application version=\"9\" xmlns=\"http://xmlns.jcp.org/xml/ns/javaee\"\n" +
+                                "    xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\n" +
+                                "    xsi:schemaLocation=\"http://xmlns.jcp.org/xml/ns/javaee\n" +
+                                "    http://xmlns.jcp.org/xml/ns/javaee/application_9.xsd\">\n" +
+                                "  <application-name>TestMessageOnStartupDefinedInEar</application-name>\n" +
+                                "  <module>\n" +
+                                "    <web>\n" +
+                                "      <web-uri>" + APP_NAME + ".war</web-uri>\n" +
+                                "      <context-root>/" + APP_NAME + "</context-root>\n" +
+                                "    </web>\n" +
+                                "  </module>\n" +
+                                "</application>";
+
+        EnterpriseArchive ear = ShrinkWrap.create(EnterpriseArchive.class, APP_NAME + ".ear")
+                        .addAsModule(war)
+                        .addAsLibrary(jar)
+                        .addAsResource(new org.jboss.shrinkwrap.api.asset.StringAsset(applicationXml), "META-INF/application.xml");
+
+        ShrinkHelper.exportAppToServer(server, ear, SERVER_ONLY, DeployOptions.DISABLE_VALIDATION);//we're testing tolerance for custom names so this needs to be manual.
+        server.addInstalledAppForValidation("TestMessageOnStartupDefinedInServerXML");
+        server.startServer();
+    }
+
+    @AfterClass
+    public static void teardown() throws Exception {
+        server.stopServer();
+    }
+
+}

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/suite/FATSuite.java
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/suite/FATSuite.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2023 IBM Corporation and others.
+ * Copyright (c) 2023, 2026 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -18,8 +18,8 @@ import io.openliberty.microprofile.reactive.messaging.fat.validation.ValidationT
 
 @RunWith(Suite.class)
 @Suite.SuiteClasses({
-    KafkaTests.class,
-    ValidationTests.class
+                      KafkaTests.class,
+                      ValidationTests.class
 })
 public class FATSuite extends TestContainerSuite {
 }

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/suite/KafkaTests.java
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/suite/KafkaTests.java
@@ -30,20 +30,22 @@ import io.openliberty.microprofile.reactive.messaging.fat.kafka.metrics.MetricsT
 import io.openliberty.microprofile.reactive.messaging.fat.kafka.metrics.MultiAppMetricsTest;
 import io.openliberty.microprofile.reactive.messaging.fat.kafka.nack.KafkaNackTest;
 import io.openliberty.microprofile.reactive.messaging.fat.kafka.validation.KafkaValidationTests;
+import io.openliberty.microprofile.reactive.messaging.fat.startup.TestMessageOnStartupTest;
 import io.openliberty.microprofile.reactive.messaging.fat.telemetry.ReactiveMessagingTelemetryTest;
 import io.openliberty.microprofile.reactive.messaging.fat.telemetry.ReactiveMessagingTelemetryTestWithJAXRS;
 
 @RunWith(Suite.class)
 @SuiteClasses({
-        KafkaNackTest.class,
-        KafkaValidationTests.class,
-        MetricsTest.class,
-        MultiAppMetricsTest.class,
-        ReactiveMessagingTelemetryTest.class,
-        ReactiveMessagingTelemetryTestWithJAXRS.class,
-        KafkaEmitterTest.class,
-        KafkaEmitterRestfulTest.class,
-        KafkaEmitterNackRestfulTest.class
+                KafkaNackTest.class,
+                KafkaValidationTests.class,
+                MetricsTest.class,
+                MultiAppMetricsTest.class,
+                ReactiveMessagingTelemetryTest.class,
+                ReactiveMessagingTelemetryTestWithJAXRS.class,
+                KafkaEmitterTest.class,
+                KafkaEmitterRestfulTest.class,
+                KafkaEmitterNackRestfulTest.class,
+                TestMessageOnStartupTest.class
 })
 public class KafkaTests {
 

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/publish/servers/TestMessageOnStartupServer/bootstrap.properties
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/publish/servers/TestMessageOnStartupServer/bootstrap.properties
@@ -1,0 +1,14 @@
+###############################################################################
+# Copyright (c) 2026 IBM Corporation and others.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License 2.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
+###############################################################################
+bootstrap.include=../testports.properties
+com.ibm.ws.logging.trace.specification = *=info:logservice=detail:REACTIVEMESSAGE=all

--- a/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/publish/servers/TestMessageOnStartupServer/server.xml
+++ b/dev/io.openliberty.microprofile.reactive.messaging.internal_fat/publish/servers/TestMessageOnStartupServer/server.xml
@@ -1,0 +1,29 @@
+<!--
+    Copyright (c) 2026 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<server>
+
+    <include location="../fatTestPorts.xml" />
+
+    <featureManager>
+        <feature>componenttest-2.0</feature>
+        <feature>localConnector-1.0</feature>
+        <feature>mpReactiveMessaging-3.0</feature>
+        <feature>enterpriseBeans-4.0</feature>
+        <feature>cdi-4.0</feature>
+        <feature>concurrent-3.0</feature>
+        <feature>mpHealth-4.0</feature>
+        <feature>servlet-5.0</feature>
+        <feature>mpConfig-3.1</feature>
+    </featureManager>
+
+    <!-- Application name defined in server.xml (different from the name in application.xml) -->
+    <application name="TestMessageOnStartupDefinedInServerXML" location="TestMessageOnStartup.ear" />
+  
+</server>

